node/src/main/scala/coop/rchain/node/proposescript/ProposeLeaderful.scala

```diff
@@ -0,0 +1,351 @@
+package coop.rchain.node.proposescript
+
+import cats.effect.Resource
+import cats.syntax.all._
+import coop.rchain.casper.protocol.{
+  BlockInfo,
+  BlockQuery,
+  BlocksQuery,
+  DeployData,
+  LightBlockInfo,
+  ProposeQuery,
+  ProposeResultQuery
+}
+import coop.rchain.casper.protocol.deploy.v1.DeployServiceV1GrpcMonix
+import coop.rchain.casper.protocol.propose.v1.ProposeServiceV1GrpcMonix
+import coop.rchain.crypto.PrivateKey
+import coop.rchain.crypto.codec.Base16
+import coop.rchain.crypto.signatures.{Secp256k1, Signed}
+import coop.rchain.shared.Log
+import io.grpc.{ManagedChannel, ManagedChannelBuilder}
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import org.rogach.scallop.ScallopConf
+import pureconfig._
+import pureconfig.generic.auto._
+import fs2._
+
+import java.nio.file.Path
+import scala.concurrent.duration.FiniteDuration
+
+final case class ProposeInfo(
+    deployContract: String,
+    phloLimit: Long,
+    phloPrice: Long,
+    deployPrivateKey: String
+)
+final case class ValidatorInfo(
+    host: String,
+    grpcPort: Int,
+    httpPort: Int,
+    proposeInfo: ProposeInfo
+) {
+  val ProposeSuccessRegex      = "Success! Block ([0-9a-f]+) created and added.".r
+  val ProposeAsyncSuccessRegex = "Propose started \\(seqNum ([0-9]+)\\)".r
+  private def startGrpcChannel: Resource[Task, ManagedChannel] =
```

Comment:
> Because with each API call the whole process will exit, it doesn't make real difference here...

What do you mean about that?

> But ManagedChannel is meant to be used as a single object for one host. It has connection pool and can be used multi threaded.

Yes. You are right. I was thinking considering channel as normal tcp connection and close it when it is done which can save some efforts of thinking higher concepts.

Reply:
What I usually do with these kind of _builders_ is to create a wrapper which accepts function that can modify builder options.
NOTE: I'm not sure if we want to copy termination logic from [DeployService](https://github.com/rchain/rchain/blob/a6457a836778/node/src/main/scala/coop/rchain/node/api/client/DeployService.scala#L183-L191).
```scala
def createChannel[F[_]: Sync](
    setup: NettyChannelBuilder => NettyChannelBuilder
): Resource[F, ManagedChannel] =
  Resource.make[F, ManagedChannel](
    Sync[F].delay(setup(NettyChannelBuilder.forAddress(host, grpcPort)).build())
  )(channel => Sync[F].delay(channel.shutdown()) >> ().pure[F])
```
Using it is easy and configurable.
```scala
val channelResource1 = createChannel[Task](identity) 
val channelResource2 = createChannel[Task](_.usePlaintext()) // with additional options
```

Reply:
> Yes, please use parameter F instead directly Task. You would also need Monixable because generated gRPC bindings uses monix Task directly.

Ok. I didn't find any way to turn Observerble into F. I got you now. I was thinking I am already at the end of the world so it is fine to be concrete. You make more sense.

Reply:
@tgrospic i still got some problems with [Observable](https://monix.io/docs/current/reactive/observable.html#observable-and-functional-programming). I don't think Observable[A] can easily turn into F[A] without introducing Task([like the old dev did]( https://github.com/rchain/rchain/blob/dev/shared/src/main/scala/coop/rchain/monix/MonixableSyntax.scala#L19)). This is one of the reason I used Task instead of F.

Reply:
Unfortunately we cannot avoid Task because how gRPC services are generated. That's why I created Monixable, to have conversion from Task to F (and reverse). So Task you get from observable can be converted to F.

Similar case is e.g. when sending a message with `TransportLayer#send` which uses Task directly.
https://github.com/rchain/rchain/blob/db297f119eded438744e57e6c7a9579db54c165b/comm/src/main/scala/coop/rchain/comm/transport/GrpcTransport.scala#L87-L102